aptos-move/framework/aptos-framework/sources/storage_gas.move

```diff
@@ -0,0 +1,359 @@
+        initialize(aptos_framework);
+    }
+
+    public(friend) fun initialize(aptos_framework: &signer) {
+        system_addresses::assert_aptos_framework(aptos_framework);
+        assert!(
+            !exists<StorageGasConfig>(@aptos_framework),
+            error::already_exists(ESTORAGE_GAS_CONFIG)
+        );
+        let item_curve = base_8192_exponential_curve(10, 10000);
+        let byte_curve = base_8192_exponential_curve(1, 1000);
+
+        let item_config = UsageGasConfig {
+            target_usage: 1000000000,  // 1 billion
+            read_curve: item_curve,
+            create_curve: item_curve,
+            write_curve: item_curve,
+        };
+        let byte_config = UsageGasConfig {
+            target_usage: 500000000000, // 500 GB
+            read_curve: byte_curve,
+            create_curve: byte_curve,
+            write_curve: byte_curve,
+        };
+        validate_usage_config(&item_config);
+        validate_usage_config(&item_config);
+        move_to(aptos_framework, StorageGasConfig {
+            item_config,
+            byte_config,
+        });
+
+        assert!(
+            !exists<StorageGas>(@aptos_framework),
+            error::already_exists(ESTORAGE_GAS)
+        );
+        move_to(aptos_framework, StorageGas {
+            per_item_read: 100,
+            per_item_create: 100,
+            per_item_write: 100,
+            per_byte_read: 1,
+            per_byte_create: 1,
+            per_byte_write: 1,
+        });
+    }
+
+    fun validate_usage_config(config: &UsageGasConfig) {
+        assert!(config.target_usage > 0, error::invalid_argument(EZERO_TARGET_USAGE));
+        assert!(config.target_usage <= 18446744073709551615 / BASIS_POINT_DENOMINATION, error::invalid_argument(ETARGET_USAGE_TOO_BIG));
+        validate_curve(&config.read_curve);
+        validate_curve(&config.create_curve);
+        validate_curve(&config.write_curve);
+    }
+
+    fun validate_curve(curve: &GasCurve) {
+        assert!(curve.max_gas >= curve.min_gas, error::invalid_argument(EINVALID_GAS_RANGE));
+        let points = &curve.points;
+        let len = vector::length(points);
+        let i = 0;
+        while (i <= len) {
+            let cur = if (i == 0) { &Point {x: 0, y: 0} } else { vector::borrow(points, i - 1) };
+            let next = if (i == len) { &Point {x: BASIS_POINT_DENOMINATION, y: BASIS_POINT_DENOMINATION} } else { vector::borrow(points, i) };
+            assert!(cur.x < next.x && cur.y <= next.y, error::invalid_argument(EINVALID_MONOTONICALLY_NON_DECREASING_CURVE));
+            i = i + 1;
+        }
+    }
+
+    fun calculate_gas(max_usage: u64, current_usage: u64, curve: &GasCurve): u64 {
+        let capped_current_usage = if (current_usage > max_usage) max_usage else current_usage;
+        let points = &curve.points;
+        let num_points = vector::length(points);
+        let current_usage_bps = capped_current_usage * BASIS_POINT_DENOMINATION / max_usage;
```

Comment:
Can capped_current_usage * BASIS_POINT_DENOMINATION overflow?

Reply:
line 186 check this.